pkg/kubernetes/auth/check_auth.go

```diff
@@ -0,0 +1,142 @@
+	switch typez {
+	case CheckNone:
+		return &NoopChecker{}, nil
+	case CheckAuthenticated:
+		return &ValidBearerTokenChecker{apiProvider: client.NewInCluster}, nil
+	case CheckAdmin:
+		return &AdminBearerTokenChecker{apiProvider: client.NewInCluster}, nil
+	}
+	return nil, fmt.Errorf("auth checker type unknown: %s. Must be one of %s, %s, %s", typez, CheckAdmin, CheckAuthenticated, CheckNone)
+}
+
+type NoopChecker struct {
+	Checker
+}
+
+func (b *NoopChecker) CheckAuth(ctx context.Context, header http.Header) error {
+	hlog.Debug("noop auth checker: ignore auth")
+	return nil
+}
+
+func getUserToken(header http.Header) (string, error) {
+	authValue := header.Get(AuthHeader)
+	if authValue != "" {
+		parts := strings.Split(authValue, "Bearer ")
+		if len(parts) != 2 {
+			return "", errors.New("missing Bearer token in Authorization header")
+		}
+		return parts[1], nil
+	}
+	return "", errors.New("missing Authorization header")
+}
+
+func runTokenReview(ctx context.Context, apiProvider client.APIProvider, token string, preds []tokenReviewPredicate) error {
+	client, err := apiProvider()
+	if err != nil {
+		return err
+	}
+
+	rvw, err := client.CreateTokenReview(ctx, &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}, &metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	for _, predFunc := range preds {
+		if err = predFunc(rvw); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type tokenReviewPredicate func(*authv1.TokenReview) error
+
+func mustBeAuthenticated(rvw *authv1.TokenReview) error {
+	if !rvw.Status.Authenticated {
+		return errors.New("user not authenticated")
+	}
+	return nil
+}
+
+func mustBeClusterAdmin(rvw *authv1.TokenReview) error {
+	for _, group := range rvw.Status.User.Groups {
+		if group == "system:cluster-admins" {
+			return nil
+		}
+	}
+	return errors.New("user not in cluster-admins group")
+}
+
+type ValidBearerTokenChecker struct {
+	Checker
+	apiProvider client.APIProvider
+}
+
+func (c *ValidBearerTokenChecker) CheckAuth(ctx context.Context, header http.Header) error {
+	hlog.Debug("Checking authenticated user")
+	token, err := getUserToken(header)
+	if err != nil {
+		return err
+	}
+	hlog.Debug("Checking auth: token found")
+	if err = runTokenReview(ctx, c.apiProvider, token, []tokenReviewPredicate{mustBeAuthenticated}); err != nil {
+		return err
+	}
+
+	hlog.Debug("Checking auth: passed")
+	return nil
+}
+
+type AdminBearerTokenChecker struct {
+	Checker
+	apiProvider client.APIProvider
+}
+
+func (c *AdminBearerTokenChecker) CheckAuth(ctx context.Context, header http.Header) error {
+	hlog.Debug("Checking authenticated user")
+	token, err := getUserToken(header)
+	if err != nil {
+		return err
+	}
+	hlog.Debug("Checking auth: token found")
+	if err = runTokenReview(ctx, c.apiProvider, token, []tokenReviewPredicate{mustBeAuthenticated, mustBeClusterAdmin}); err != nil {
+		return err
+	}
+
+	hlog.Debug("Checking auth: passed")
+	return nil
+}
```

Comment:
I guess this code was duplicated for future improvments but at the end I don't see the value to do it. 
`[]tokenReviewPredicate` could be passed as a simple argument with the `apiProvider` in `NewChecker` function. Is it going to be other differences than that ?

Reply:
I started with a slightly different implementation between the two checkers but after one or two iterations it ended up being almost the same, modulo predicates .. so yeah we can merge them into a single impl; if they diverge again later then we would split them back

Reply:
:+1: perfect ! thanks